r/R/query-engine.R

```diff
@@ -81,8 +82,8 @@ ExecPlan <- R6Class("ExecPlan",
           # head and tail are not ExecNodes; at best we can handle them via
           # SinkNode, so if there are any steps done after head/tail, we need to
           # evaluate the query up to then and then do a new query for the rest.
-          # as_record_batch_reader() will build and run an ExecPlan
-          node <- self$SourceNode(as_record_batch_reader(.data$.data))
+          # as_arrow_table() will build and run an ExecPlan
+          node <- self$SourceNode(as_arrow_table(.data$.data))
```

Comment:
Yeah this would force you to consume the whole RBR to take the head, no bueno. Can you detect whether a UDF is used in `.data$.data` and only do `as_arrow_table` in that case? (Maybe even just whether there are any UDFs registered is close enough?)

(Side note: this line will need special handling in #13541 too, it's a sneaky place where `$Build()` actually starts running the query, which we don't want there.)

Reply:
Can you make a followup JIRA for this? I think there are a few options, but agree that we should just error for now, no need to grow the scope even further at this point.